Modularization

ukstv opened the pull request:
Divide the package into linking (light) and validation (heavy) subpackages.

Beware:
- pnpm: npm and lerna suck,
- subpackages: breaks CircleCI and automated code coverage reports

ukstv committed:
feat: linking package

ukstv committed:
feat: validation package

ukstv committed:
chore: add test snapshots under github control

oed commented:
I like this structure generally. It's clean and keeps the validation part as one complete package. The main problem with this is that we use it in the `@ceramicnetworkdoctype-caip10-link` package in js-ceramic which is also used in the `@ceramicnetwork/http-client` which we want to keep down the size of. However, I think it might be possible to have the http-client only depend on the doctype and not the handler, which would remove this issue altogether.

Would love some more indepth thoughts on the changes @zachferland @PaulLeCam 

What do we think about pnpm?

oed commented:
Here is a PR in ceramic that could potentially mitigate the issue mentioned above: https://github.com/ceramicnetwork/js-ceramic/pull/668

ukstv commented:
Naaa-a-ah! Found the solution. Easier, but a bit weird.

PaulLeCam commented on README.md:
````diff
@@ -1,84 +1,145 @@
 ## Install
+
 ```
 $ npm install --save 3id-blockchain-utils
````
This should change no?

PaulLeCam commented on packages/validation/package.json:
```diff
@@ -0,0 +1,40 @@
+  },
+  "author": "oed",
+  "license": "(Apache-2.0 OR MIT)",
+  "dependencies": {
```
Shall we split this package into one package for each blockchain so we can make the dependencies more manageable?

ukstv commented:
This: #47 solves the trouble. Runs all right in browser with no zondax installed.

The same trick could be done to Polkadot and EOSio in order to reduce the resulting web package.

PaulLeCam commented on README.md:
```diff
@@ -1,84 +1,145 @@
+- [validate](#validation) the link.
+
+Due to asymmetry in dependencies - linking is light weigth, while validation requires quite heavy code - and target use cases,
+the package is divided into two: namely `@3id-blockchain-utils/linking` and `@3id-blockchain-utils/validation`.
```
Is this the namespace we want to use rather than `@ceramicnetwork`?

oed replied:
We should use `@ceramicnetwork/blockchain-linking` and `@ceramicnetwork/blockchain-validation`

PaulLeCam reviewed (commented):
LGTM but I'm not familiar with the specifics so my comments are just about the setup.

Maybe we should discuss the use of pnpm a bit more, I've never used it before so I don't know the pros and cons compared to alternatives.
Overall though it would be good to unify the tools we're using. npm v7 supports workspaces so maybe it's something we could try out first to avoid adding pnpm to our tools unless strictly necessary?

oed reviewed (approved):
I think this looks like a good start!

ukstv committed:
chore: packages use @ceramicnetwork name scope

ukstv committed:
chore: packages use @ceramicnetwork name scope

ukstv committed:
fix: use @ceramicnetwork name scope for deps in tests

ukstv committed:
chore: bring back eos in validation

ukstv committed:
fix: use truly dynamic import for zondax signing tools

ukstv committed:
chore: strict node versioning

ukstv committed:
chore: make tests run

Really, Jest tries to use fancy Node API that is experimental in Node v14. So, we enable this.

ukstv commented:
@oed Here it is.
